### PR TITLE
new class _Cached()

### PR DESCRIPTION
**Pull request by mcepl**
* clean names
* use template method for stamp
* lru is list (not iterator)

Fails in CI https://travis-ci.org/mcepl/rope/builds/111960979

### CODE REVIEW

File: rope/base/utils.py [hunk 1]
```diff
@@ -57,72 +59,94 @@ def newfunc(*args, **kwds):
...
+def to_hashable(obj):
+    """
+    Makes a hashable object from a dictionary, list, tuple, set etc.
+    """
+    if isinstance(obj, (list, tuple)):
+        return tuple(to_hashable(i) for i in obj)
+    elif isinstance(obj, (set, frozenset)):
+        return frozenset(to_hashable(i) for i in obj)
+    elif isinstance(obj, dict):
+        new_obj = {k: to_hashable(v) for k, v in obj.items()}
```
Comment (mcepl) [inline rope/base/utils.py]: This doesn’t work in ``2.6``:
```python
>>> a = range(40)
>>> {(x, x**2) for x in a}
  File "<stdin>", line 1
    {(x, x**2) for x in a}
                 ^
SyntaxError: invalid syntax
```
However, this works:
```python
set((x, x**2) for x in a)
```

File: rope/base/utils.py [hunk 2]
```diff
@@ -1,5 +1,7 @@
+import random
 import warnings
-from functools import partial
+
+from time import time
```
Comment (mcepl) [inline rope/base/utils.py]: I would prefer ``import time`` and use ``time.time()`` then. I slightly dislike ``from ... import ...`` style of imports as they tend to make code less readable (especially with common words like ``time``).

Comment (emacsway) [inline rope/base/utils.py]: > I would prefer import time and use time.time() then.

There is special case (cache should be as fast as possible), see [Avoiding dots...](https://wiki.python.org/moin/PythonSpeed/PerformanceTips#Avoiding_dots...)

> Hmm, and time actually doesn't seem to be used at all.

I've excluded time in next commit, simple counter faster in 2 times than time().

Comment (mcepl) [inline rope/base/utils.py]: Could you please rebase yourself your branch https://github.com/emacsway/rope/tree/cache_optimization on top of my https://github.com/mcepl/rope/tree/sublimeText_speedup ? I am doing something wrong. Then make a new pull request to that branch of mine (or just let me know, I can do it myself, if you don't know how).

**Pull request closed without merging**